File: Day-9: Recursion/3) Combination sum-1.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n = candidates.size();
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> check = {};
        helper(0, target, candidates, ans, check);
        return ans;
    }
    void helper(int id, int target, vector<int> &nums, vector<vector<int>> &ans, vector<int> &check)
    {
        int n = nums.size();
        if(target == 0)
        {
            ans.push_back(check);
            return;
        }
        for(int i = id; i < n; i++)
        {
            if(nums[i] > target)
            {
                break;
            }
            check.push_back(nums[i]);
            helper(i, target - nums[i], nums, ans, check);
            check.pop_back();
        }
    }
};

File: Day-9: Recursion/3) Combination sum-1.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> check = new ArrayList<>();
        helper(0, target, candidates, ans, check);
        return ans;
    }
    private void helper(int index, int target, int[] nums, List<List<Integer>> ans, List<Integer> check)
    {
        if(target == 0)
        {
            ans.add(new ArrayList<>(check));
            return;
        }
        for(int i = index; i < nums.length; i++)
        {
            if(nums[i] > target)
            {
                break;
            }
            check.add(nums[i]);
            helper(i, target - nums[i], nums, ans, check);
            check.remove(check.size() - 1);
        }
    }
}
